### ISSUE TITLE
Allow `user` and other client properties to be set after starting crash service

### ISSUE THREAD

**Issue by ndonald2** [getsentry/sentry-swift#110]
### Overview

Properties on `SentryClient` which are set after calling `startCrashHandler` are not sent along with crash reports. This includes `user` as well as `tags`. I have not verified this for `extra`. 

It makes sense to begin crash handling as soon as possible in the app lifecycle, but at that point user information or other tag data may not be available, for example if the user has not logged in yet, or if user info is not kept in cache and is instead downloaded from the server on every app session. However, setting the `user` or other properties after starting crash handling has no apparent effect.

### Expected Result

Setting `user`, `tags`, etc on the shared `SentryClient` instance at any point during the app's lifecycle, including after already starting crash handling, should cause that information to be sent with crash reports for any subsequent crashes.

### Actual Result

No custom client data is sent with crash reports if set after starting crash handling.

### Example

This works correctly:

```objective-c
SentryClient *client = [[SentryClient alloc] initWithDsnString:@"<DSN>"];
[SentryClient setShared:client];
SentryUser *sentryUser = [[SentryUser alloc] initWithId:@"1"
                                                  email:@"testuser@fake.com"
                                               username:@"Claudia Thompson"
                                                  extra:@{}];
[[SentryClient shared] setUser:sentryUser];
[[SentryClient shared] startCrashHandler];
```

This does not work correctly:

```objective-c
SentryClient *client = [[SentryClient alloc] initWithDsnString:@"<DSN>"];
[SentryClient setShared:client];
[[SentryClient shared] startCrashHandler];

// Later
SentryUser *sentryUser = [[SentryUser alloc] initWithId:@"1"
                                                  email:@"testuser@fake.com"
                                               username:@"Claudia Thompson"
                                                  extra:@{}];
[[SentryClient shared] setUser:sentryUser];
```

**Comment by ndonald2** [comment]
Looking through the source, it seems like user info isn't persisted to disk anywhere in either `Sentry` or `KSCrash`. Setting `user` on `SentryClient` after crash reporting is started results in the user info getting set on `crashHandler` since it will not be nil at that point.

```swift
public var user: User? = nil {
    didSet { crashHandler?.user = user }
}
```

However, by then any enqueued crash reports will have already been submitted, since that happens immediately when calling `startCrashHandler`.

```swift

// SentryClient.swift 

internal(set) var crashHandler: CrashHandler? {
    didSet {
        crashHandler?.startCrashReporting()
        crashHandler?.releaseVersion = releaseVersion
        crashHandler?.tags = tags
        crashHandler?.extra = extra
        crashHandler?.user = user
    }
}

// KSCrashHandler.swift

internal func startCrashReporting() {
    // Sychrnoizes this function
    objc_sync_enter(lock)
    defer { objc_sync_exit(lock) }
    
    // Return out if already installed
    if isInstalled { return }
    isInstalled = true
    
    // Install
    installation.install()
    
    KSCrash.sharedInstance().searchThreadNames = true
    
    Log.Debug.log("Started Sentry Client \(SentryClient.versionString)")
    
    sendAllReports()
}
```

I'm actually confused as to how the user info even gets through if set before calling `startCrashHandler`. In the `didSet` closure for `crashHandler`, the reports are sent before the user info is updated. Assuming the reports are formatted and sent asynchronously, wouldn't that at best be a race condition with updating the user info for the report format?

 In other words, shouldn't `crashHandler?.startCrashReporting()` be the _last_ line of `didSet`?

**Comment by HazAT** [comment]
@ndonald2
Hey thanks for that detailed bug description ... I am currently looking into that.

**Comment by HazAT** [comment]
@ndonald2 
I am not able to reproduce it, also created a test for it ... could you take a look at it and tell me if I didn't understand it correctly? 
https://github.com/getsentry/sentry-swift/pull/111

**Comment by ndonald2** [comment]
@HazAT 
Thanks for looking into it and writing the test. Unfortunately the test does not represent the scenario I was describing. 

In the event of an actual crash, all application state and memory contents will be destroyed. So even if the `user` had been configured during the application's lifetime, it will need to be reconfigured on the next application launch. However, as soon as `startCrashHandler` is called, any pending crash reports will be formatted and uploaded. So if the user has not yet been configured, the report will not be associated with a user.

Here's an outline of the timing which I hope illustrates the problem:

* App Launch
* Start Crash Handling
* Set User on `SentryClient`
* App Crashes
* App Opened Again
* Start Crash Handling
    * Pending Crash Report Sent (**No User Data Set**)
* Set User on `SentryClient` (**Too Late**)

 There is nothing in the documentation that states that you must set `user` prior to calling `startCrashHandling` in order for crash reports to properly be associated with a user, so I would expect it to work either way. Granted, many (if not most) apps will already have the necessary information at hand to setup the `user` when the application first launches, i.e. if the user information/object is persisted to disk, but this is not the case for all apps, including mine - the user's profile is retrieved from the API on every app launch.

Because reports are formatted and sent immediately on the next app launch when calling `startCrashHandling`, it seems that the only way to fix the issue as described would be for `SentrySwift` to persist `user` (and `tags`, etc) to disk and then load that data when `SentryClient` is reinitialized on the next app launch.

Please let me know if that does not make sense. If this is not a use case that `SentrySwift` will support, then I would request that the documentation be updated in order to indicate that `user` must be set prior to calling `startCrashHandling` in order to have a user associated with a crash report. Thanks!

**Comment by HazAT** [comment]
@ndonald2 
Thanks for the detailed description, now I fully understand the problem.
I've recently discovered that behaviour when using breadcrumbs ... and yes it should work that way that in case of a crash, a snapshot of all set properties should be persisted and stored and sent with these properties on the next startup.
I am gonna definitely fix that, stay tuned.

**Comment by HazAT** [comment]
The new release `1.4.3` fixes this issue, thanks again for pointing that out.

**Comment by ndonald2** [comment]
Great, thank you!
